`init` event fires before leaving empty state

Consider the following model definition:

``` javascript
import Model from "ember-data/model";
import attr from "ember-data/attr";
import { belongsTo, hasMany } from "ember-data/relationships";
import Ember from 'ember';

export default Model.extend({
  title: attr(),
  propSetFromObservingTitle: attr(),

  setPropFromTitle: Ember.observer('title', function() {
      this.set('propSetFromObservingTitle', this.get('title') + 'hi');
  }).on('init')
});
```

In this example, we have two properties that are `DS.attr`'s. We have `propSetFromObservingTitle` because we basically want a computed property that _also_ gets serialized to the server.

This code works with normal ember objects, but when we run the following code:

``` javascript
this.store.createRecord('post', {title: 'foo'});
```

We get the following exception:

```
VM423 ember.debug.js:31321 Error: Attempted to handle event `didSetProperty` on <post:null> while in state root.empty. Called with {name: propSetFromObservingTitle, oldValue: undefined, originalValue: undefined, value: undefinedhi}.
    at new Error (native)
    at Error.EmberError (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:15976:21)
    at InternalModel._unhandledEvent (https://cdnjs.cloudflare.com/ajax/libs/ember-data.js/2.7.0/ember-data.js:2623:13)
    at InternalModel.send (https://cdnjs.cloudflare.com/ajax/libs/ember-data.js/2.7.0/ember-data.js:2501:14)
    at Class.set (https://cdnjs.cloudflare.com/ajax/libs/ember-data.js/2.7.0/ember-data.js:13361:31)
    at Descriptor.computedPropertySet [as _set] (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:14654:28)
    at Descriptor.computedPropertySetWithSuspend [as setWithSuspend] (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:14634:19)
    at Descriptor.computedPropertySetEntry [as set] (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:14612:17)
    at Object.set (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:20348:12)
    at Class.set (https://cdnjs.cloudflare.com/ajax/libs/ember.js/2.6.2/ember.debug.js:34672:38)
```

Example on Ember Twiddle: https://ember-twiddle.com/415538f55bbaa01ca95067ec0e890d43?openFiles=models.post.js%2C

--- a/addon/-private/system/model/internal-model.js
+++ b/addon/-private/system/model/internal-model.js
@@ -319,7 +319,7 @@ export default class InternalModel {
     return this.currentState.dirtyType;
   }
 
-  getRecord() {
+  getRecord(properties) {
     if (!this._record && !this._isDematerializing) {
       heimdall.increment(materializeRecord);
       let token = heimdall.start('InternalModel.getRecord');
@@ -334,6 +334,10 @@ export default class InternalModel {
         isError: this.isError,
         adapterError: this.error
       };
+
+      if (typeof properties === 'object' && properties !== null) {
+        assign(createOptions, properties);
+      }
 
       if (setOwner) {
         // ensure that `getOwner(this)` works inside a model instance

--- a/addon/-private/system/store.js
+++ b/addon/-private/system/store.js
@@ -365,16 +365,8 @@ Store = Service.extend({
     properties.id = coerceId(properties.id);
 
     let internalModel = this._buildInternalModel(normalizedModelName, properties.id);
-    let record = internalModel.getRecord();
-
-    // Move the record out of its initial `empty` state into
-    // the `loaded` state.
-    // TODO @runspired this seems really bad, store should not be changing the state
     internalModel.loadedData();
-
-    // Set the properties specified on the record.
-    // TODO @runspired this is probably why we do the bad thing above
-    record.setProperties(properties);
+    let record = internalModel.getRecord(properties);
 
     // TODO @runspired this should also be coalesced into some form of internalModel.setState()
     internalModel.eachRelationship((key, descriptor) => {

--- a/tests/unit/model-test.js
+++ b/tests/unit/model-test.js
@@ -1329,6 +1329,27 @@ test('toJSON looks up the JSONSerializer using the store instead of using JSONSe
   assert.deepEqual(json, {});
 });
 
+test('internalModel is ready by `init`', function(assert) {
+  assert.expect(2);
+  let nameDidChange = 0;
+
+  const Person = DS.Model.extend({
+    name: DS.attr('string'),
+
+    init() {
+      this._super(...arguments);
+      this.set('name', 'my-name-set-in-init');
+    },
+
+    nameDidChange: Ember.observer('name', () => nameDidChange++)
+  });
+
+  let { store } = setupStore({ person: Person });
+
+  assert.equal(nameDidChange, 0, 'observer should not trigger on create');
+  let person = run(() => store.createRecord('person'));
+  assert.equal(person.get('name'), 'my-name-set-in-init');
+});
 
 test('accessing attributes in the initializer should not throw an error', function(assert) {
   assert.expect(1);

